[MCH] initial implementation of workflows

aferrero2707 opened the pull request:
This PR introduces a first implementation of the MCH DPL workflows for the decoding and pre-clustering steps.

The introduced workflows are the following:
* Raw data input from file saved by readout.exe: 
`o2-mch-file-to-digits-workflow --infile [file name] --frames [number of HBF to read]`
* Raw data decoding: `o2-mch-raw-to-digits-workflow`
* Digits output: `o2-mch-digits-sink-workflow --outfile [file name]`
* Pre-clustering: `o2-mch-digits-to-preclusters-workflow`
* Pre-clusters output: `o2-mch-preclusters-sink-workflow --outfile [file name]`

aferrero2707 commented:
@aphecetche @pillot could you review this PR?

aphecetche commented:
@aferrero2707 your file-reader.cfg example mention only a ped file (thus a BareFormat file). I assume this means the workflows are not supposed to work with UL data for the moment ?

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+
+///
+/// \file    runFileReader.cxx
+/// \author  Andrea Ferrero
```
well, I guess this comment is from a cut & paste and not completely fitted to this file ?

aferrero2707 replied:
Fixed.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+#include "MCHRawCommon/RDHManip.h"
+
+#include "MCHWorkflow/DataDecoderSpec.h"
+//#include "RawBufferDecoder.h"
```
remove

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+#include "MCHWorkflow/DataDecoderSpec.h"
+//#include "RawBufferDecoder.h"
+
+// Dans ce code, on récupère un infut aui est un message avec le buffer, on fait tourner le code de base decodeBuffer qui est dans Handlers, et on renvoir un message de sortie (inspiré de FileReader de Andrea)
```
remove (or translate to English ;-) )

aferrero2707 replied:
Removed.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+    size_t ndigits{0};
+
+    auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+      auto s = asString(dsElecId);
```
for performance reasons, this debug string should only be computed where needed (in the mPrint if below)

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+        //Segmentation segment(deId);
+
+        padId = segment.findPadByFEE(dsIddet, int(channel));
+        if(mPrint)
```
I guess in the not-to-distant future we should either a) remove those debug statement b) use the proper Logger functions instead of bare cout)

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+      int padId = -1;
+      try {
+        const Segmentation& segment = segmentation(deId);
+        //Segmentation segment(deId);
```
remove commented code

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+      int time = 0;
+
+      digits.emplace_back( o2::mch::Digit(time, deId, padId, digitadc) );
+      //o2::mch::Digit& mchdigit = digits.back();
```
remove commented code

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+  }
+
+private:
+  std::function<std::optional<DsDetId>(DsElecId)> Elec2Det;
```
class members variable names should start with m (see https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html#Variable_Names)

aferrero2707 replied:
Ok. I also removed the un-needed fee2Solar member.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -0,0 +1,261 @@
+{
+  return DataProcessorSpec{
+    "DataDecoder",
+    //Inputs{InputSpec{"readout", "ROUT", "RAWDATA", Lifetime::Timeframe}},
```
remove commented code

aferrero2707 replied:
Done. I also removed the processing of the "readout" input in the `run()` function.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/PreClusterFinderSpec.cxx:
```diff
@@ -32,7 +32,7 @@
 #include "MCHBase/Digit.h"
 #include "MCHBase/PreCluster.h"
-#include "PreClusterFinder.h"
+#include "../../PreClustering/src/PreClusterFinder.h"
```
No, that's not the correct way to reference includes from another target. If you need the PreClusterFinder.h and it's not currently exported by the MCHPreClustering target, then that must be changed. @pillot any objection to export that file ?

aphecetche replied:
@pillot Actually I'm starting to look at this and it might requires a bit of work if we don't want to expose the PreClusterFinderMapping (and I guess we don't want to).

aphecetche replied:
@pillot I made an attempt to "hide" the PreClusterFinderMapping while still exporting the PreClusterFinder.h, see https://github.com/AliceMCH/AliceO2/compare/mch-dpl-workflows...aphecetche:expose-preclusterfinder?expand=1 
I've just tried the compilation stage though (i.e. the easy part ;-) ), as the (closure) tests we've already discussed about are not currently in AliceO2, so it'd be good if you could try and see if I broke everything or not with this "structural" change...

pillot replied:
Thanks for the solution! It looks fine but I will run some tests next week.

aferrero2707 replied:
What should we do for the moment? I agree that the above include statement is an ugly workaround...

aphecetche commented on Detectors/MUON/MCH/Workflow/src/digits-reader-tb-workflow.cxx:
```diff
@@ -0,0 +1,45 @@
+/// \file    runFileReader.cxx
+/// \author  Andrea Ferrero
+///
+/// \brief This is an executable that reads a data file from disk and sends the data to QC via DPL.
```
same remark as above : update to reflect current file

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/digits-reader-tb-workflow.cxx:
```diff
@@ -0,0 +1,45 @@
+#include "Framework/runDataProcessing.h"
+#include "TBDigitsFileReaderSpec.h"
+
+// Dans ce code, on récupère un infut aui est un message avec le buffer, on fait tourner le code de base decodeBuffer qui est dans Handlers, et on renvoir un message de sortie (inspiré de FileReader de Andrea)
```
same as above : translate or remove

aferrero2707 replied:
Removed.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/digits-sink-workflow.cxx:
```diff
@@ -0,0 +1,143 @@
+// or submit itself to any jurisdiction.
+
+///
+/// \file    runFileReader.cxx
```
update to reflect current file

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:
```diff
@@ -0,0 +1,314 @@
+/// \file    runFileReader.cxx
+/// \author  Andrea Ferrero
+///
+/// \brief This is an executable that reads a data file from disk and sends the data to QC via DPL.
```
see above

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:
```diff
@@ -0,0 +1,314 @@
+
+class FileReaderTask
+{
+  void decodeBuffer(gsl::span<const std::byte> page, std::vector<o2::mch::Digit> &digits)
```
this looks a lot like the decodeBuffer in DataDecoderSpec, can't this be made a bit more DRY ?

aferrero2707 replied:
Yes, I will try to put the common part into a separate source file.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/raw-to-digits-workflow.cxx:
```diff
@@ -0,0 +1,47 @@
+///
+/// \brief This is an executable that reads a data file from disk and sends the data to QC via DPL.
+///
+/// This is an executable that reads a data file from disk and sends the data to QC via the Data Processing Layer.
```
same as above : rewrite

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/raw-to-digits-workflow.cxx:
```diff
@@ -0,0 +1,47 @@
+#include "Framework/runDataProcessing.h"
+#include "MCHWorkflow/DataDecoderSpec.h"
+
+// Dans ce code, on récupère un infut aui est un message avec le buffer, on fait tourner le code de base decodeBuffer qui est dans Handlers, et on renvoir un message de sortie (inspiré de FileReader de Andrea)
```
same as above : translate or remove

aferrero2707 replied:
Removed.

aphecetche commented:
@aferrero2707 I've made a number of comments above. 
I'd have one more about the TB (test beam) files. I wonder whether those files should go in AliceO2 really or be kept elsewhere (e.g. in https://github.com/mrrtf/alo ?) as they're not meant to reconstruct Run3 data at all

pillot commented on Detectors/MUON/MCH/Workflow/src/TBDigitsFileReader.h:
```diff
@@ -0,0 +1,93 @@
+#include "MCHBase/Digit.h"
+
+using namespace o2::mch;
+using namespace std;
```
same comment about using namespace in headers

aferrero2707 replied:
Removed.

pillot commented on Detectors/MUON/MCH/Workflow/src/TBDigitsFileReaderSpec.cxx:
```diff
@@ -0,0 +1,91 @@
+/// \file TBDigitsFileReaderSpec.cxx
+/// \brief Implementation of a data processor to run the preclusterizer
+///
+/// \author Philippe Pillot, Subatech
```
I guess you copy/past these comments ;-)

aferrero2707 replied:
Yes, I corrected it.

pillot commented on Detectors/MUON/MCH/Workflow/src/TBDigitsFileReaderSpec.h:
```diff
@@ -0,0 +1,34 @@
+#include "Framework/DataProcessorSpec.h"
+
+using namespace o2;
+using namespace o2::framework;
```
same comment about using namespace in headers

aferrero2707 replied:
Removed.

pillot commented on Detectors/MUON/MCH/Workflow/src/digits-reader-run2-workflow.cxx:
```diff
@@ -0,0 +1,35 @@
+/// \file PreClusterFinderDevice.cxx
+/// \brief Implementation of a DPL device to run the preclusterizer
+///
+/// \author Philippe Pillot, Subatech
```
same as above

aferrero2707 replied:
Corrected.

pillot commented on Detectors/MUON/MCH/Workflow/src/digits-to-preclusters-workflow.cxx:
```diff
@@ -0,0 +1,35 @@
+/// \file PreClusterFinderDevice.cxx
+/// \brief Implementation of a DPL device to run the preclusterizer
+///
+/// \author Philippe Pillot, Subatech
```
same as above

aferrero2707 replied:
Corrected.

pillot commented on Detectors/MUON/MCH/Workflow/src/preclusters-sink-workflow.cxx:
```diff
@@ -0,0 +1,35 @@
+/// \file PreClusterFinderDevice.cxx
+/// \brief Implementation of a DPL device to run the preclusterizer
+///
+/// \author Philippe Pillot, Subatech
```
same as above

aferrero2707 replied:
Corrected.

pillot commented on Detectors/MUON/MCH/Workflow/src/preclusters-sink-workflow.cxx:
```diff
@@ -0,0 +1,35 @@
+using namespace o2;
+using namespace o2::framework;
+
+// clang-format off
```
why disabling the clang format?

aferrero2707 replied:
Removed.

pillot commented on Detectors/MUON/MCH/Workflow/src/TBDigitsFileReaderSpec.h:
```diff
@@ -0,0 +1,34 @@
+/// \file PreClusterFinderSpec.h
+/// \brief Definition of a data processor to run the preclusterizer
+///
+/// \author Philippe Pillot, Subatech
```
same a above

aferrero2707 replied:
Corrected.

pillot commented on Detectors/MUON/MCH/Workflow/src/digits-reader-run2-workflow.cxx:
```diff
@@ -0,0 +1,35 @@
+using namespace o2;
+using namespace o2::framework;
+
+// clang-format off
```
why disabling clang format?

aferrero2707 replied:
This is some copy-paste left-over...

pillot commented on Detectors/MUON/MCH/Workflow/src/raw-to-digits-workflow.cxx:
```diff
@@ -0,0 +1,47 @@
+using namespace o2;
+using namespace o2::framework;
+
+// clang-format off
```
why disabling clang format?

aferrero2707 replied:
See above.

pillot commented on Detectors/MUON/MCH/Workflow/src/digits-reader-tb-workflow.cxx:
```diff
@@ -0,0 +1,45 @@
+using namespace o2;
+using namespace o2::framework;
+
+// clang-format off
```
why disabling clang format?

aferrero2707 replied:
See above.

pillot commented on Detectors/MUON/MCH/Workflow/src/digits-sink-workflow.cxx:
```diff
@@ -0,0 +1,130 @@
+#include "MCHBase/Digit.h"
+
+using namespace o2;
+using namespace o2::framework;
```
this is repeated below

aferrero2707 replied:
Removed.

pillot commented on Detectors/MUON/MCH/Workflow/src/digits-to-preclusters-workflow.cxx:
```diff
@@ -0,0 +1,35 @@
+using namespace o2;
+using namespace o2::framework;
+
+// clang-format off
```
why disabling clang-format?

aferrero2707 replied:
Removed.

aferrero2707 commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoderSpec.h:
```diff
@@ -0,0 +1,22 @@
+
+using namespace o2;
+using namespace o2::framework;
+
```
Removed.

aferrero2707 commented on Detectors/MUON/MCH/Workflow/src/PreClusterFinderSpec.cxx:
```diff
@@ -109,13 +109,6 @@ class PreClusterFinderTask
-        precluster.print(cout, mUsedDigits);
-      }
-    }
-
```
Done.

aferrero2707 commented on Detectors/MUON/MCH/Workflow/src/PreClusterSinkSpec.cxx:
```diff
@@ -95,6 +103,14 @@ class PreClusterSinkTask
       mOutputFile.write(reinterpret_cast<const char*>(digits.data()), digits.size_bytes());
     }
+
+    bool mPrint = false;
```
Done.

aferrero2707 commented:
@aphecetche @pillot I have included most of the suggested improvements.
The remaining open questions are:
* how to deal with the `PreClusterFinder.h` include file
* combine the common code in `DataDecoderSpec.cxx` and `file-to-digits-workflow.cxx` into a separate source file
* move the sources that handle the test-beam and Run2 data files into a separate repository (but which one?)

I still have to re-run the clang formatting on the sources.

aphecetche commented:
@aferrero2707 for the preclustering the idea would be to merge in this PR the changes from https://github.com/aphecetche/AliceO2/pull/4 that makes the PreClusterFinder.h include exported.

For the handling of Run2 (and test beam) data I would put things into https://github.com/mrrtf/alo which was meant for that kind of stuff.

pillot commented:
I tested the new workflows, piping the digit sampler, preclusterizer and precluster sink, and it works fine. I get the same results and timing. Therefore I think we can (completely) remove the original workflow. I made a PR to this branch: https://github.com/AliceMCH/AliceO2/pull/5

pillot commented:
> For the handling of Run2 (and test beam) data I would put things into https://github.com/mrrtf/alo which was meant for that kind of stuff.

I think there is a misunderstanding. Probably the name of the workflow should be digits-reader-workflow instead of digits-reader-run2-workflow. The digit sampler works with run3 data, with an option to use it for "run2 data" (in which case the PADId contains the run2 DigitId, which is converted into PADId by the spec). But the data format is still run3 digit. The reason I did all this is exactly to be able to run the full chain 100% in O2.

aferrero2707 commented:
@aphecetche @pillot I have included the code from the [expose-preclusterfinder](https://github.com/aphecetche/AliceO2/tree/expose-preclusterfinder) branch, and merged https://github.com/AliceMCH/AliceO2/pull/5. I have also removed the test-beam data reader from this PR, and removed the `run2` suffix from the digits reader workflow file name.

I have also added a `--txt` option to the digits and preclusters sink workflows, allowing to switch from binary to human-readable text output.

In order to test the digits sink and reader workflows, I have done the following:
```
o2-mch-file-to-digits-workflow -b --infile ../Data/data-flp-20200324.raw --frames 10 | o2-mch-digits-sink-workflow -b --outfile digits.out
o2-mch-file-to-digits-workflow -b --infile ../Data/data-flp-20200324.raw --frames 10 | o2-mch-digits-sink-workflow -b --outfile digits.txt -txt
o2-mch-digits-reader-workflow --infile digits.out -b | o2-mch-digits-sink-workflow -b --outfile digits2.txt --txt
```
and I checked that `digits.txt` and `digits2.txt` have the same content.
`../Data/data-flp-20200324.raw` is a pedestals data file collected at P2 from CH5.

pillot commented:
Thanks @aferrero2707 , everything looks fine on my side.

aphecetche commented:
> > For the handling of Run2 (and test beam) data I would put things into https://github.com/mrrtf/alo which was meant for that kind of stuff.
> 
> I think there is a misunderstanding. Probably the name of the workflow should be digits-reader-workflow instead of digits-reader-run2-workflow. The digit sampler works with run3 data, with an option to use it for "run2 data" (in which case the PADId contains the run2 DigitId, which is converted into PADId by the spec). But the data format is still run3 digit. The reason I did all this is exactly to be able to run the full chain 100% in O2.

@pillot Thanks for the clarification. But I guess my comment still stands for the Test Beam stuff, right ?

sawenzel commented:
It looks like the CI is stuck on this PR.

sawenzel commented:
There are some genuine code checker violations:
```
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/PreClustering/src/PreClusterFinder.cxx:49:19: error: use '= default' to define a trivial destructor [modernize-use-equals-default]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:192:23: error: use nullptr [modernize-use-nullptr]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:200:5: error: field reference 'nFrames' does not match naming rule [aliceO2-member-name]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:228:9: error: field reference 'nFrames' does not match naming rule [aliceO2-member-name]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:231:9: error: field reference 'nFrames' does not match naming rule [aliceO2-member-name]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:232:7: error: field reference 'nFrames' does not match naming rule [aliceO2-member-name]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:266:23: error: use nullptr [modernize-use-nullptr]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/MUON/MCH/Workflow/src/file-to-digits-workflow.cxx:282:7: error: field declaration 'nFrames' does not match naming rule [aliceO2-member-name]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0//Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoderSpec.h:1:1: error: missing or malformed copyright notice
```

We can go on once they are fixed.

sawenzel commented:
@ktf @MichaelLettrich : The CI did not report anything here although the last commit was 18h ago. Is there any problem?

MichaelLettrich commented:
HI, not exactly sure why this takes so long, the macOS builders work fine as far as I can see from the logs. the last commit that was built for this PR is b1cfc82. It seems that we simply have a high number of open PRs - If by chance some packages had to be rebuilt because of changes in alidist, you soon end up with a very high load on the CI.

MichaelLettrich commented:
@aferrero2707 can you force-push this one, so that it gets picked up by the CI again? I honestly don't know why the CI ignores it.

aferrero2707 commented:
@MichaelLettrich how do I force-push? I have tried running `git push --force` on my branch but nothing happened on the CI side...

ktf commented:
You are a first time contributor. This requires an explicit approval as described here:

https://aliceo2group.github.io/quickstart/devel.html#pull-request-flow

I will approve you.

MichaelLettrich commented:
Oh this is new for me as well @ktf, thanks for letting me know. 
@aferrero2707 for the record:
In case someone tells you to "force push" or "bump" your repo for the CI you need to alter the hash of the commit for example by adding a newline somewhere, commit and push it. 
If you think adding a commit which just adds a newline is not nice in your history, you can rebase afterwards, squash that commit into a previous commit and then force-push the branch.

aferrero2707 commented:
@MichaelLettrich Ha! Then it was simpler than I thought...

aferrero2707 commented:
@ktf @MichaelLettrich there is only one failing check (`build/O2/gpu`), but looking at the logs I have the impression that the failure is unrelated with this PR...

MichaelLettrich commented:
@aferrero2707  yes this is unrelated. merging.

davidrohr commented:
This is weird, fails to compile for me because gsl::span doesn't know length. Shouldn't it be size?
The below patch fixes it for me. Anyone has similar problems?
```
diff --git a/Detectors/MUON/MCH/Workflow/src/digits-sink-workflow.cxx b/Detectors/MUON/MCH/Workflow/src/digits-sink-workflow.cxx
index a1e6180b107..26d9bff1822 100644
--- a/Detectors/MUON/MCH/Workflow/src/digits-sink-workflow.cxx
+++ b/Detectors/MUON/MCH/Workflow/src/digits-sink-workflow.cxx
@@ -88,7 +88,7 @@ class DigitsSinkTask
         mOutputFile << " DE# " << d.getDetID() << " PadId " << d.getPadID() << " ADC " << d.getADC() << " time " << d.getTimeStamp() << std::endl;
       }
     } else {
-      int nDigits = digits.length();
+      int nDigits = digits.size();
       mOutputFile.write(reinterpret_cast<char*>(&nDigits), sizeof(int));
       mOutputFile.write(reinterpret_cast<const char*>(digits.data()), digits.size_bytes());
     }
```

aphecetche commented:
@davidrohr seems you are right but none of the compilers caught it so far (I guess because that part of the code is a branch that currently cannot be taken ?). What's your configuration ?

aphecetche commented:
@aferrero2707 yes, David found the reason ("old" ms_gsl used in AliceO2), see https://github.com/AliceO2Group/AliceO2/pull/3414
